### PR TITLE
refactor apiGateway resources and methods

### LINKED ISSUE #1738: HTTP Events, cannot add methods for root `/`endpoint
##### Serverless Framework Version:

1.0.0-alpha.2
##### Operating System:

OSX 10.11.6
##### Expected Behavior:

I would expect that I would be able to somehow deploy a method against the root `/` endpoint. Seems like validations block this from happening? Either that or I missed in the docs how to handle this.
##### Actual Behavior:

``` yaml
functions:
  render:
    handler: handler.render
    timeout: 6
    events:
      - http:
          path: /
          method: get
```

```
♪ sls deploy
Serverless: Zipping service...
Serverless: Removing old service versions...
Serverless: Uploading .zip file to S3...
Serverless: Updating Stack...

  Validation Error ---------------------------------------

     Template format error: Unresolved resource dependencies
     [ResourceApigEvent] in the Resources block of the template
```

### LINKED ISSUE #2025: Refactor method.js file for cleaner code
# This is a Feature Proposal
## Description

methods.js (`lib/plugins/aws/deploy/compile/events/apiGateway/lib/methods.js`) in the APIG Event is way too complicated at this point and needs to be refactored to run in separate functions or even separate modules/files to have cleaner code. This will also make future feature implementations a lot easier

Similar or dependent issues:
- 
## Additional Data
- **_Serverless Framework Version you're using**_:
- **_Operating System**_:
- **_Stack Trace**_:
- **_Provider Error messages**_:

### LINKED ISSUE #2124: Can't deploy very simple project, Unresolved resource dependencies [ApiGatewayResource]
# This is a (Bug Report / Feature Proposal)

Bug Report
## Description

I have an extremely minimal test serverless.yml nearly unchanged from the template:

```
service: oden

provider:
  name: aws
  runtime: python2.7
  stage: dev
  region: us-west-1

functions:
  crash:
    handler: handler.log_crash
    events:
      - http:
          path: /crash
          method: post
```

when I try to deploy via `sls deploy -v` I get the following message:

```
     Template format error: Unresolved resource dependencies
     [ApiGatewayResource] in the Resources block of the template
```

The first time I tried this it printed a bit of extra information before giving the same error:

```
CloudFormation - CREATE_IN_PROGRESS - AWS::CloudFormation::Stack - oden-dev
CloudFormation - CREATE_IN_PROGRESS - AWS::S3::Bucket - ServerlessDeploymentBucket
CloudFormation - CREATE_IN_PROGRESS - AWS::S3::Bucket - ServerlessDeploymentBucket
CloudFormation - CREATE_COMPLETE - AWS::S3::Bucket - ServerlessDeploymentBucket
CloudFormation - CREATE_COMPLETE - AWS::CloudFormation::Stack - oden-dev
```

but on subsequent deploys it no longer prints those messages. I tried to deploy with `--noDeploy` to look at the CloudFormation stack files; here is the portion of `cloudformation-template-update-stack.json` where it mentions `ApiGatewayResource`:

```
    "ApiGatewayResourceCrash": {
      "Type": "AWS::ApiGateway::Resource",
      "Properties": {
        "ParentId": {
          "Ref": "ApiGatewayResource"
        },
        "PathPart": "crash",
        "RestApiId": {
          "Ref": "ApiGatewayRestApi"
        }
      }
    },
```
## Additional Data
- **_Serverless Framework Version you're using**_:
  v1.0.0-rc1
- **_Operating System**_:
  Mac OS X 
- **_Stack Trace**_:
- **_Provider Error messages**_:

### LINKED ISSUE #2188: Refactor tests in API Gateways method.js file
# This is a Feature Proposal
## Description

The `method.js` test file is a little bit bloated and many tests rely on configuration which is defined at the top of the file. This makes it hard to write new tests / refactor or update old tests.

Every test should re-define it's own `http` event definition so that they live near the test itself and won't collide / introduce surprises as something slips through.

There's already a cleanup task in the making?! Would be good if this is referenced here...

### PR DESCRIPTION
**Pull request by dougmoscrop**
## What did you implement:

Closes #2025 
Closes #1738
Closes #2124
Closes #2188

## How did you implement it:

This is the last of my previous pull request, broken up. It touches resources and methods to completely move apiGateway compilation to be consistent (all objects, no more string parse/unparse). It makes the tests independently configured. It fixes up the issues with using root resources.

## How can we verify it:

- `npm test` passes
- `npm run simple-integration-test` passes
- `npm run complex-integration-test` passes

## Todos:

- [x] Write tests
- [x] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config/commands/resources
- [x] Change ready for review message below

***Is this ready for review?:*** YES

### CODE REVIEW

**Comment by dougmoscrop** [comment]
Sorry hold off on the review, I've messed up a rebase.

**Comment by pmuens** [comment]
Hey @dougmoscrop thanks for opening this up and working on this! 👍  🌴  

Really looking forward to get this in as well. The API Gateway refactoring you've done so far is rock solid 🚀 ! Great job!

**Comment by dougmoscrop** [comment]
Ah! Sorry it's so big still; it's moving a lot of code around.

I can split resources off in to another PR, but it's just the changes in it don't necessarily make sense without methods support.. 

I'll open another PR with resources

**Comment by flomotlik** [comment]
@dougmoscrop is the rebase done now?

@pmuens @eahefnawy  do you feel comfortable reviewing this? Imho this is a large PR but if we can get it in quickly we got a lot of refactoring done now (thanks @dougmoscrop btw)

**Comment by dougmoscrop** [comment]
Almost; some CORS stuff needs to be fixed. Working on it.

**Comment by dougmoscrop** [comment]
Ok, @flomotlik , @pmuens good to review. All tests (including integration) passed after I fixed the rebase mistake I made.

**Comment by eahefnawy** [comment]
@dougmoscrop you're a champ! thanks a lot! ☺️  looking into this now ... we want to get this in asap to avoid any future conflicts

Great changes 👍

File: lib/plugins/aws/deploy/compile/events/apiGateway/lib/methods.js [hunk 1]
```diff
@@ -3,658 +3,47 @@
...
+        this.getMethodIntegration(event.http, event.functionName),
+        this.getMethodResponses(event.http)
+      );
 
-        if (responseConfig.integrationResponseTemplate) {
-          _.merge(integrationResponse.ResponseTemplates, {
-            'application/json': responseConfig.integrationResponseTemplate,
-          });
-        }
+      const methodName = event.http.method[0].toUpperCase() +
```
Comment (pmuens) [inline lib/plugins/aws/deploy/compile/events/apiGateway/lib/methods.js]: Could we use the `_.upperFirst()` method here as well?

File: lib/plugins/aws/deploy/compile/events/apiGateway/lib/method/integration.js [hunk 2]
```diff
@@ -0,0 +1,194 @@
+'use strict';
+
+const _ = require('lodash');
+
+module.exports = {
+  getMethodIntegration(http, functionName) {
+    const normalizedFunctionName = functionName[0].toUpperCase()
```
Comment (pmuens) [inline lib/plugins/aws/deploy/compile/events/apiGateway/lib/method/integration.js]: Could we use the `_.upperFirst()` method here as well?

File: lib/plugins/aws/deploy/compile/events/apiGateway/lib/method/authorization.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
+'use strict';
+
+module.exports = {
+  getMethodAuthorization(http) {
+    if (http.authorizer) {
+      const normalizedAuthorizerName = http.authorizer.name[0].toUpperCase()
```
Comment (pmuens) [inline lib/plugins/aws/deploy/compile/events/apiGateway/lib/method/authorization.js]: Could we use the `_.upperFirst()` method here as well?

File: lib/plugins/aws/deploy/compile/events/apiGateway/lib/validate.js [hunk 4]
```diff
@@ -151,6 +238,146 @@ module.exports = {
...
+        ' Please check the docs for more info.',
+      ].join('');
+      throw new this.serverless.classes.Error(errorMessage);
+    }
+    return http.response;
+  },
+
   getLambdaArn(name) {
     this.serverless.service.getFunction(name);
     const normalizedName = name[0].toUpperCase() + name.substr(1);
```
Comment (pmuens) [inline lib/plugins/aws/deploy/compile/events/apiGateway/lib/validate.js]: Could we use the `_.upperFirst()` method here as well?

**Comment by pmuens** [comment]
Hey @dougmoscrop thank you so much for this 🎉  👍 !

We've moved your changes into a new PR (https://github.com/serverless/serverless/pull/2666) so that we can apply some minor updates. Closing this one because of that...

Enjoy your vacation! 🏕  ⛺️

**Pull request closed without merging**